### PR TITLE
DID and JWK hardening

### PR DESCRIPTION
**Pull request by logangirvin**
**Problem:**
Defense in depth: add additional verifications to resolved DID documents and specifically their JWKs to bind them to the requested identifiers.

**Solution:**
Additional verifications.

**Validation:**
Unit tests

**Type of change:**
- [ ] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [x] Test
- [ ] Logging/Telemetry

**Risk**:
- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [x] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

**Work Item links**:

**Documentation Links**:

### CODE REVIEW

File: WalletLibrary/WalletLibrary/StatusCheck/StatusListTokenValidator.swift [hunk 1]
```diff
@@ -107,18 +111,17 @@ class StatusListTokenValidator {
...
+        guard let keyIdentifier = DIDVerificationMethodIdentifier(keyId: keyId) else {
+            throw StatusListValidationError.malformedKeyId
+        }
 
-        let matchingKeys = keys.filter { $0.id == keyId || (keyFragment != nil && $0.id == keyFragment) }
-        for key in matchingKeys where verify(token, with: key) {
+        if (try? token.verify(
+            using: tokenVerifier,
+            keys: keys,
+            keyIdentifier: keyIdentifier)) == true {
```
Comment (josephpab) [inline WalletLibrary/WalletLibrary/StatusCheck/StatusListTokenValidator.swift]: **Issue:** the `try?` now wraps the *whole key loop* instead of a single key. Previously `verifySignature` filtered matching keys and used a per-key `(try? token.verify(using:withPublicKey:)) == true`, so a key that **threw** (unsupported curve/alg, malformed JWK) was skipped and the next matching key was still tried. `JwsToken.verify(using:keys:keyIdentifier:)` propagates on the first throwing key, and `try?` here collapses that to `nil` for the entire loop.

**Impact:** when a document has more than one key matching the `kid`, verification now fails as soon as the *first* candidate throws, even if a later key would have verified — the caller then degrades the credential to `.unknown`. That is a behavior regression versus `dev` and is unrelated to the DID-binding hardening this PR is adding. The other four call sites are unaffected (they already used bare `try`), so this is the only place the semantics changed.

**Recommendation:** keep the new binding but restore per-key tolerance:

```swift
for key in keys where key.matches(keyIdentifier) {
    if (try? token.verify(using: tokenVerifier, withPublicKey: key.publicKeyJwk)) == true {
        return
    }
}

throw StatusListValidationError.invalidSignature
```

`Severity: Medium`

File: WalletLibrary/WalletLibrary/Extensions/Entities/IdentifierDocument+WalletLibrary.swift [hunk 2]
```diff
@@ -9,21 +9,21 @@
  */
 extension IdentifierDocument
 {
-    func getJWK(id: String) -> JWK?
+    func getJWK(id: String, forDID did: String? = nil) -> JWK?
```
Comment (josephpab) [inline WalletLibrary/WalletLibrary/Extensions/Entities/IdentifierDocument+WalletLibrary.swift]: **Issue:** `forDID` defaults to `nil`, and the `nil` branch (line 20) is exactly the pre-hardening lookup this PR sets out to remove — `publicKeys.first(where: { $0.id == id })`, with no controller check and no DID binding.

**Impact:** the unvalidated path is still the **default**. After this change the only production caller is `SignedCredentialMetadataProcessor`, which does pass `forDID`; the sole remaining users of the default are the three assertions in `IdentifierDocumentExtensionTests`. Any future call site that omits the argument silently opts out of key binding with no compiler or review signal — which is how the gap being closed here appeared in the first place.

**Recommendation:** make the parameter required — `func getJWK(id: String, forDID did: String) -> JWK?` — and update the three tests to pass the document's own DID. The compiler then enforces the binding instead of relying on callers to remember it.

`Severity: Medium`

File: WalletLibrary/Submodules/VerifiableCredential-SDK-iOS/VCNetworking/VCNetworking/operations/fetch/FetchDIDDocumentOperation.swift [hunk 3]
```diff
@@ -3,6 +3,76 @@
...
+            throw malformedInput("Resolved Identifier Document does not match the requested DID.")
+        }
+    }
+
+    private static func validate(identifier: String) throws {
+        let allowedCharacters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._-%:"
+        guard !identifier.isEmpty,
+              identifier.allSatisfy({ allowedCharacters.contains($0) }) else {
+            throw malformedInput("Invalid DID identifier.")
+        }
```
Comment (josephpab) [inline WalletLibrary/Submodules/VerifiableCredential-SDK-iOS/VCNetworking/VCNetworking/operations/fetch/FetchDIDDocumentOperation.swift]: **Issue:** `allowedCharacters` is applied only to the raw `identifier`. Each decoded `candidate` produced by the loop below is checked solely by `validateDecodedIdentifier`, which rejects `/ \ ? #` and `.`/`..` segments but never re-applies the allowlist. Percent escapes that decode to characters the allowlist would have rejected outright therefore pass — `%20` -> space, `%00` -> NUL, `%40` -> `@`, `%3B` -> `;`.

**Impact:** low in practice. `URLComponents.path` re-encodes `%` as `%25` before the request is sent — I confirmed `did:web:example.com:%2e%2e:evil` goes out as `.../did:web:example.com:%252e%252e:evil` — so the decoded form is never what reaches the discovery service. But the entire purpose of this decode loop is to reason about what a server sees after *its* decoding, and this is the one hole left in an otherwise strict allowlist.

**Recommendation:** hoist `allowedCharacters` (minus `%`) into `validateDecodedIdentifier` and apply it to every candidate, so decoded forms are held to the same character policy as the raw input.

`Severity: Low`

File: WalletLibrary/Submodules/VerifiableCredential-SDK-iOS/VCNetworking/VCNetworkingTests/operations/fetch/FetchDIDDocumentOperationTests.swift [hunk 4]
```diff
@@ -28,4 +28,58 @@ class FetchDIDDocumentOperationTests: XCTestCase {
...
+            "did:web:",
+            "did:web:example.com::evil",
+            "did:web:example.com:.:evil",
+            "did:web:example.com:..:evil",
+            "did:web:example.com:%2e%2e:evil",
+            "did:web:example.com:%252e%252e:evil",
+            "did:web:example.com:%252525252e%252525252e:evil",
+            "did:web:example.com%2Fevil",
+            "did:web:example.com%5Cevil",
+            "did:web:example.com%3Fversion=1",
```
Comment (josephpab) [inline WalletLibrary/Submodules/VerifiableCredential-SDK-iOS/VCNetworking/VCNetworkingTests/operations/fetch/FetchDIDDocumentOperationTests.swift:54]: **Issue:** this vector and `"did:web:example.com?version=1"` on line 58 never reach the query-string / percent-decoding logic they are meant to cover. `=` is absent from `allowedCharacters`, so both are rejected by the raw character allowlist on the very first guard in `validate(identifier:)`.

**Impact:** two of the 18 vectors assert the wrong thing. If the `?`/`#` rejection inside `validateDecodedIdentifier` regressed, these two would still pass — false confidence on precisely the traversal class this PR is hardening.

**Recommendation:** drop the `=` so the intended guard is exercised, e.g. `"did:web:example.com%3Fversion"` and `"did:web:example.com?version"`. More broadly, `XCTAssertThrowsError` here only asserts *that* something threw; consider asserting the rejection reason so a vector cannot quietly shift to a different guard.

`Severity: Low`

File: WalletLibrary/Submodules/VerifiableCredential-SDK-iOS/VCEntities/VCEntities/identifier/document/IdentifierDocumentPublicKey.swift [hunk 5]
```diff
@@ -25,4 +25,10 @@ struct IdentifierDocumentPublicKey: Codable, Equatable {
     init(fromJwk key: PublicJWK) {
         self.init(id: key.keyId, type: VCEntitiesConstants.SUPPORTED_PUBLICKEY_TYPE, controller: nil, publicKeyJwk: key, purposes: [VCEntitiesConstants.PUBLICKEY_AUTHENTICATION_PURPOSE_V1])
     }
+
+    func matches(_ keyIdentifier: DIDVerificationMethodIdentifier) -> Bool {
+        let controllerMatches = controller == nil || controller == keyIdentifier.did
+        let identifierMatches = id == keyIdentifier.absoluteId || id == keyIdentifier.relativeId
+        return controllerMatches && identifierMatches
```
Comment (josephpab) [inline WalletLibrary/Submodules/VerifiableCredential-SDK-iOS/VCEntities/VCEntities/identifier/document/IdentifierDocumentPublicKey.swift:32]: **Issue:** `matches(_:)` and `DIDVerificationMethodIdentifier` are now the single choke point for key binding across five call sites — `PresentationRequestValidator`, `IssuanceRequestValidator`, `DomainLinkageCredentialValidator`, `StatusListTokenValidator` and `IdentifierDocument.getJWK` — yet neither type has a direct unit test. Coverage is entirely indirect, via the three new `PresentationRequestValidatorTests` cases.

**Impact:** the security-relevant edges are unpinned. Note that `controller == nil` makes `controllerMatches` unconditionally `true`, so a bare `#fragment` key id is accepted on identifier match alone. That is only safe because this PR *also* added the `document.id == requestedIdentifier` check upstream — nothing in the suite documents or protects that coupling, so a future refactor that relaxes the document check would silently re-open cross-DID key acceptance. `getJWK(id:forDID:)` gained no new tests at all.

**Recommendation:** add a dedicated test file covering absolute id match, absolute id belonging to a different DID, relative id with `nil` controller, relative id with a mismatched controller, and — for `DIDVerificationMethodIdentifier` — multi-`#`, leading `#`, trailing `#`, and no-`#` inputs.

`Severity: Low`

**Review by josephpab — APPROVED** [review]

File: WalletLibrary/Submodules/VerifiableCredential-SDK-iOS/VCNetworking/VCNetworkingTests/operations/fetch/FetchDIDDocumentOperationTests.swift (hunk 4 above)
Comment (logangirvin) [reply WalletLibrary/Submodules/VerifiableCredential-SDK-iOS/VCNetworking/VCNetworkingTests/operations/fetch/FetchDIDDocumentOperationTests.swift:54]: While I agree its not great to have test cases that fail for different reasons, we have a LOT of test cases here so in this case I think its okay.
